rfcs/react-components/styles-handbook.md

```diff
@@ -0,0 +1,815 @@
+
+## `makeStyles`
+
+`makeStyles` is used to define style permutations in components and is used for style overrides. Returns a [React hook][react-hook] that should be called inside a component:
+
+```js
+import { makeStyles } from '@griffel/react';
+
+const useClasses = makeStyles({
+  button: { display: 'flex' },
+  icon: { paddingLeft: '5px' },
+});
+```
+
+### Limitations
+
+`makeStyles()` does not support [CSS shorthands][griffel-css-shorthands-support] in styles definitions. However, Griffel provides a set of [`shorthands` functions][griffel-css-shorthands] to mimic them:
+
+```js
+import { makeStyles, shorthands } from '@griffel/react';
+
+const useClasses = makeStyles({
+  root: {
+    // ❌ This is not supported, TypeScript compiler will throw, styles will not be inserted to DOM
+    padding: '2px 4px 8px 16px',
+    // ✅ Use shorthand functions to avoid writting CSS longhands
+    ...shorthands.padding('2px', '4px', '8px', '16px'),
+  },
+});
+```
+
+> 💡 **Note:** The most of the functions follow syntax in matching CSS properties, but each value should a separate argument:
+
+```js
+// ❌ Will produce wrong results:
+//   {
+//     paddingBottom: "2px 4px"
+//     paddingLeft: "2px 4px"
+//     paddingRight: "2px 4px"
+//     paddingTop: "2px 4px"
+//   }
+shorthands.padding('2px 4px');
+// ✅ Correct output:
+//   {
+//     paddingBottom: "2px"
+//     paddingLeft: "4px"
+//     paddingRight: "4px"
+//     paddingTop: "2px"
+//   }
+shorthands.padding('2px', '4px');
+```
+
+### Performance caveat
+
+Atomic CSS generates more classes than a standard approach with monolithic classes. Usually it's not a problem, but there are cases when performance [may degrade][griffel-recalc-performance], for example when elements have more than 100 classes. This does not happen usually as there are only a limited number of CSS properties that can be applied to a DOM element:
+
+```js
+import { makeStyles, shorthands } from '@griffel/react';
+
+const useClasses = makeStyles({
+  root: {
+    display: 'flex',
+    alignItems: 'center',
+    ...shorthands.padding('4px'),
+  },
+});
+// ⬇️⬇️⬇️
+// produces 6 classes (1+1+4)
+```
+
+> 💡**Tip:** Preview generated classes consider in the [Try out][griffel-try-out] sandbox.
+
+However, "CSS rule explosion" can happen when using nested selectors, [pseudo classes][mdn-pseudo-classes]/selectors and [At-rules][mdn-at-rules]:
+
+```js
+import { makeStyles, shorthands } from '@griffel/react';
+
+const useClasses = makeStyles({
+  root: {
+    ...shorthands.padding('4px'),
+    ...shorthands.margin('4px'),
+
+    ':hover': {
+      ...shorthands.padding('4px'),
+      ...shorthands.margin('4px'),
+    },
+
+    '::before': {
+      display: 'block',
+      content: "' '",
+    },
+
+    '@media (forced-colors: active)': {
+      ...shorthands.padding('4px'),
+      ...shorthands.margin('4px'),
+    },
+  },
+});
+// ⬇️⬇️⬇️
+// produces 14 classes ((4+4)+(4+4)+2+(4+4))
```

Comment:
```suggestion
// produces 26 classes ((4+4)+(4+4)+2+(4+4))
```

@miroslavstastny eagle eye 🦅 Initially the example was simpler, but I decided to highlight the problem more.